### PR TITLE
[contracts] Add getPartialAmountCeil and isRoundingErrorCeil

### PR DESCRIPTION
**Pull request by Recmo**
## Description

**Includes:** #1005.

* Adds `getPartialAmountCeil` and `isRoundingErrorCeil` that round up instead of down.
* Add `DIVISION_BY_ZERO` error to new and existing functions.
* Update/add test coverage.

## Testing instructions

Added test coverages. Added tests have merge conflicts with PR #1000.

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist:

*   [x] Prefix PR title with `[WIP]` if necessary.
*   [x] Prefix PR title with bracketed package name(s) corresponding to the changed package(s). For example: `[sol-cov] Fixed bug`.
*   [x] Add tests to cover changes as needed.
*   [x] Update documentation as needed.
*   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/contracts/src/2.0.0/protocol/Exchange/libs/LibMath.sol [hunk 1]
```diff
@@ -29,7 +29,7 @@ contract LibMath is
     /// @param numerator Numerator.
     /// @param denominator Denominator.
     /// @param target Value to calculate partial of.
-    /// @return Partial value of target.
+    /// @return Partial value of target rounded down.
```
Comment (hysz) [inline packages/contracts/src/2.0.0/protocol/Exchange/libs/LibMath.sol]: If we change this to `getPartialAmountFloor` then it forces developers to consider the direction of rounding.

File: packages/contracts/src/2.0.0/protocol/Exchange/libs/LibMath.sol [hunk 2]
```diff
@@ -60,16 +88,67 @@ contract LibMath is
...
+        // Otherwise, we want the relative rounding error to be strictly
+        // less than 0.1%.
+        // The relative error is `remainder / numerator * target`.
+        // We want the relative error less than 1 / 1000:
+        //        remainder / numerator * denominator  <  1 / 1000
+        // or equivalently:
+        //        1000 * remainder  <  numerator * target
+        // so we have a rounding error iff:
+        //        1000 * remainder  >=  numerator * target
         uint256 remainder = mulmod(target, numerator, denominator);
```
Comment (hysz) [inline packages/contracts/src/2.0.0/protocol/Exchange/libs/LibMath.sol]: Really nice explanation 👍

File: packages/contracts/src/2.0.0/protocol/Exchange/libs/LibMath.sol [hunk 3]
```diff
@@ -60,16 +88,67 @@ contract LibMath is
...
+        // be zero. Fortunately, this is always the case. The ideal value is
+        // zero iff `numerator == 0` and/or `target == 0`. In this case the
+        // remainder and absolute error are also zero. 
+        if (target == 0 || numerator == 0) {
+            return false;
+        }
+        
+        // Otherwise, we want the relative rounding error to be strictly
+        // less than 0.1%.
+        // The relative error is `remainder / numerator * target`.
```
Comment (hysz) [inline packages/contracts/src/2.0.0/protocol/Exchange/libs/LibMath.sol]: nit - `remainder / (numerator * target)`

File: packages/contracts/src/2.0.0/protocol/Exchange/libs/LibMath.sol [hunk 4]
```diff
@@ -60,16 +88,67 @@ contract LibMath is
...
+        // We want the relative error less than 1 / 1000:
+        //        remainder / numerator * denominator  <  1 / 1000
```
Comment (hysz) [inline packages/contracts/src/2.0.0/protocol/Exchange/libs/LibMath.sol]: `remainder / (numerator * target) < 1 / 1000`

File: packages/contracts/test/exchange/internal.ts [hunk 5]
```diff
@@ -193,18 +199,55 @@ describe('Exchange core internal functions', () => {
...
+    });
+
+    describe('getPartialAmountCeil', async () => {
+        async function referenceGetPartialAmountCeilAsync(
+            numerator: BigNumber,
+            denominator: BigNumber,
+            target: BigNumber,
+        ) {
+            if (denominator.eq(0)) {
+                throw new Error('revert DIVISION_BY_ZERO');
```
Comment (Recmo) [inline packages/contracts/test/exchange/internal.ts]: Use `throw divisionByZeroErrorForCall`

File: packages/contracts/test/exchange/internal.ts [hunk 6]
```diff
@@ -193,18 +199,55 @@ describe('Exchange core internal functions', () => {
...
+            [uint256Values, uint256Values, uint256Values],
+        );
+    });
+
+    describe('getPartialAmountCeil', async () => {
+        async function referenceGetPartialAmountCeilAsync(
+            numerator: BigNumber,
+            denominator: BigNumber,
+            target: BigNumber,
+        ) {
```
Comment (abandeali1) [inline packages/contracts/test/exchange/internal.ts]: Missing return types here.

**Pull request merged**